File: ai-pet-commerce/backend/app/services/subscription_service.py
from sqlalchemy.orm import Session
from ..models.database import User, AnalyticsRecord
from datetime import datetime, timezone

PLANS = {
    "free": {
        "name": "免费版",
        "price": 0,
        "limits": {
            "daily_selections": 5,
            "max_products": 50,
            "ai_analysis": True,
            "auto_publish": False,
            "analytics": False
        }
    },
    "pro": {
        "name": "专业版",
        "price": 99,
        "limits": {
            "daily_selections": 50,
            "max_products": 500,
            "ai_analysis": True,
            "auto_publish": True,
            "analytics": True
        }
    },
    "business": {
        "name": "企业版",
        "price": 299,
        "limits": {
            "daily_selections": 200,
            "max_products": 5000,
            "ai_analysis": True,
            "auto_publish": True,
            "analytics": True
        }
    }
}

def get_plan_limits(plan: str):
    return PLANS.get(plan, PLANS["free"])

def upgrade_subscription(db: Session, user_id: int, plan: str) -> bool:
    if plan not in PLANS:
        return False
    
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        return False
    
    user.subscription_plan = plan
    user.subscription_status = "active"
    user.updated_at = datetime.now(timezone.utc)
    
    db.commit()
    return True

def downgrade_subscription(db: Session, user_id: int) -> bool:
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        return False
    
    user.subscription_plan = "free"
    user.subscription_status = "inactive"
    user.stripe_subscription_id = None
    user.updated_at = datetime.now(timezone.utc)
    
    db.commit()
    return True

def check_limit(db: Session, user_id: int, limit_type: str) -> bool:
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        return False
    
    plan = get_plan_limits(user.subscription_plan)
    limits = plan["limits"]
    
    if limit_type == "selections":
        today = datetime.now(timezone.utc).date()
        today_selections = db.query(AnalyticsRecord)\
            .filter(AnalyticsRecord.user_id == user_id)\
            .filter(AnalyticsRecord.metric_type == "selection")\
            .filter(AnalyticsRecord.record_date >= datetime.combine(today, datetime.min.time()))\
            .count()
        return today_selections < limits["daily_selections"]
    
    elif limit_type == "products":
        product_count = db.query(User).filter(User.id == user_id).first().products.count()
        return product_count < limits["max_products"]
    
    return True

def record_analytics(db: Session, user_id: int, metric_type: str, value: float = 1.0):
    record = AnalyticsRecord(
        user_id=user_id,
        metric_type=metric_type,
        value=value,
        record_date=datetime.now(timezone.utc)
    )
    db.add(record)
    db.commit()

File: server/src/services/subscriptionService.ts
import { prisma } from '../lib/prisma.js';

const EXEMPT_ACCOUNTS = [
  'testuser@example.com',
  'demo@aipet.com',
  'admin@aipet.com'
];

export function isExemptAccount(email: string): boolean {
  return EXEMPT_ACCOUNTS.includes(email.toLowerCase());
}

export interface PlanLimits {
  dailySelections: number;
  dailyGenerations: number;
  maxProducts: number;
  aiAnalysis: boolean;
  autoPublish: boolean;
  analytics: boolean;
  apiAccess: boolean;
}

export interface Plan {
  id: string;
  name: string;
  price: number;
  currency: string;
  billingPeriod: string;
  limits: PlanLimits;
  features: string[];
}

export const PLANS: Record<string, Plan> = {
  free: {
    id: 'free',
    name: '免费版',
    price: 0,
    currency: 'CNY',
    billingPeriod: 'month',
    limits: {
        dailySelections: 5,
        dailyGenerations: 3,
        maxProducts: 50,
        aiAnalysis: true,
        autoPublish: false,
        analytics: false,
        apiAccess: false
      },
      features: [
        '每日5次AI选品',
        '每日3次内容生成',
        '最多50个商品',
        '基础AI分析',
        '社区支持'
      ]
  },
  pro: {
    id: 'pro',
    name: '专业版',
    price: 29,
    currency: 'CNY',
    billingPeriod: 'month',
    limits: {
      dailySelections: 1000,
      dailyGenerations: 1000,
      maxProducts: 500,
      aiAnalysis: true,
      autoPublish: true,
      analytics: true,
      apiAccess: true
    },
    features: [
      '无限AI分析',
      '无限内容生成',
      '最多500个商品',
      '高级AI分析',
      '数据报表分析',
      '一键上架多平台',
      'API接口访问',
      '优先技术支持'
    ]
  },
  enterprise: {
    id: 'enterprise',
    name: '企业版',
    price: 299,
    currency: 'CNY',
    billingPeriod: 'year',
    limits: {
      dailySelections: 10000,
      dailyGenerations: 10000,
      maxProducts: 5000,
      aiAnalysis: true,
      autoPublish: true,
      analytics: true,
      apiAccess: true
    },
    features: [
      '全部Pro功能',
      '批量选品',
      'API接口',
      '数据导出',
      '团队账号',
      '专属客户经理',
      '定制化功能开发',
      'SLA服务保障'
    ]
  }
};

export async function getPlan(planId: string): Promise<Plan> {
  return PLANS[planId] || PLANS.free;
}

export async function upgradeSubscription(userId: number, planId: string): Promise<boolean> {
  const plan = PLANS[planId];
  if (!plan) {
    throw new Error('无效的套餐');
  }

  const days = planId === 'pro' ? 30 : planId === 'enterprise' ? 365 : 0;
  const expireTime = days > 0 ? new Date(Date.now() + days * 24 * 60 * 60 * 1000) : null;

  await prisma.user.update({
    where: { id: userId },
    data: {
      plan: planId,
      expireTime,
      updatedAt: new Date()
    }
  });

  return true;
}

export async function downgradeSubscription(userId: number): Promise<boolean> {
  await prisma.user.update({
    where: { id: userId },
    data: {
      plan: 'free',
      expireTime: null,
      updatedAt: new Date()
    }
  });

  return true;
}

export async function checkDailyLimit(userId: number, limitType: 'selections' | 'generations'): Promise<boolean> {
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) return false;

  const plan = PLANS[user.plan];
  const today = new Date();
  today.setHours(0, 0, 0, 0);

  const count = await prisma.analyticsRecord.count({
    where: {
      userId,
      metricType: limitType,
      recordDate: { gte: today }
    }
  });

  if (limitType === 'selections') {
    return count < plan.limits.dailySelections;
  }
  return count < plan.limits.dailyGenerations;
}

export async function recordAnalytics(userId: number, metricType: string, value: number = 1): Promise<void> {
  await prisma.analyticsRecord.create({
    data: {
      userId,
      metricType,
      value,
      recordDate: new Date()
    }
  });
}

export async function getSubscriptionStatus(userId: number) {
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) return null;

  return {
    plan: user.plan,
    expireTime: user.expireTime,
    details: PLANS[user.plan]
  };
}

export async function getDailyUsage(userId: number): Promise<{ selections: number; generations: number; planLimits: PlanLimits }> {
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) {
    return {
      selections: 0,
      generations: 0,
      planLimits: PLANS.free.limits
    };
  }

  const plan = PLANS[user.plan];
  const today = new Date();
  today.setHours(0, 0, 0, 0);

  const [selectionCount, generationCount] = await Promise.all([
    prisma.analyticsRecord.count({
      where: {
        userId,
        metricType: 'selections',
        recordDate: { gte: today }
      }
    }),
    prisma.analyticsRecord.count({
      where: {
        userId,
        metricType: 'generations',
        recordDate: { gte: today }
      }
    })
  ]);

  return {
    selections: selectionCount,
    generations: generationCount,
    planLimits: plan.limits
  };
}